File: Sources/Sentry/SentryDispatchQueueWrapper.m
#import "SentryDispatchQueueWrapper.h"
#import "SentryLog.h"
#import "SentryThreadWrapper.h"
#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@interface SentryDispatchQueueWrapper ()

@property (nonatomic, assign) BOOL callingThread;

@end

@implementation SentryDispatchQueueWrapper

- (instancetype)init
{
    // DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL is requires iOS 10. Since we are targeting
    // iOS 9 we need to manually add the autoreleasepool.
    dispatch_queue_attr_t attributes = dispatch_queue_attr_make_with_qos_class(
        DISPATCH_QUEUE_SERIAL, DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    self = [self initWithName:"sentry-default" attributes:attributes];
    return self;
}

- (instancetype)initWithName:(const char *)name
                  attributes:(nullable dispatch_queue_attr_t)attributes;
{
    if (self = [super init]) {
        _queue = dispatch_queue_create(name, attributes);
        _callingThread = NO;
    }
    return self;
}

- (void)dispatchAllOnCallingThread
{
    SENTRY_LOG_DEBUG(@"dispatchAllOnCallingThread: YES");
    _callingThread = YES;
}

- (void)dispatchAsyncWithBlock:(void (^)(void))block
{
    if (self.callingThread == YES) {
        SENTRY_LOG_DEBUG(@"CallingThread: Yes");
        block();
        return;
    }

    dispatch_async(_queue, ^{
        @autoreleasepool {
            block();
        }
    });
}

- (void)dispatchAsyncOnMainQueue:(void (^)(void))block
{
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_async(dispatch_get_main_queue(), ^{
            @autoreleasepool {
                block();
            }
        });
    }
}

- (void)dispatchSync:(void (^)(void))block
{
    dispatch_sync(_queue, block);
}

- (void)dispatchSyncOnMainQueue:(void (^)(void))block
{
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_sync(dispatch_get_main_queue(), block);
    }
}

- (BOOL)dispatchSyncOnMainQueue:(void (^)(void))block timeout:(NSTimeInterval)timeout
{
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);

        dispatch_async(dispatch_get_main_queue(), ^{
            block();
            dispatch_semaphore_signal(semaphore);
        });

        dispatch_time_t timeout_t
            = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(timeout * NSEC_PER_SEC));
        return dispatch_semaphore_wait(semaphore, timeout_t) == 0;
    }
    return YES;
}

- (void)dispatchAfter:(NSTimeInterval)interval block:(dispatch_block_t)block
{
    dispatch_time_t delta = (int64_t)(interval * NSEC_PER_SEC);
    dispatch_time_t when = dispatch_time(DISPATCH_TIME_NOW, delta);
    dispatch_after(when, _queue, ^{
        @autoreleasepool {
            block();
        }
    });
}

- (void)dispatchCancel:(dispatch_block_t)block
{
    dispatch_block_cancel(block);
}

- (void)dispatchOnce:(dispatch_once_t *)predicate block:(void (^)(void))block
{
    dispatch_once(predicate, block);
}

- (nullable dispatch_block_t)createDispatchBlock:(void (^)(void))block
{
    return dispatch_block_create(0, block);
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Core/Helper/SentryDispatchQueueWrapper.swift
import Foundation

/**
 * A wrapper around DispatchQueue functions for testability.
 */
@_spi(Private) @objc public class SentryDispatchQueueWrapper: NSObject {
    
    @objc public let queue: DispatchQueue
    
    @objc public var usesDispatchOnce: Bool {
        true
    }

    @objc public override init() {
        // DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL is requires iOS 10. Since we are targeting
        // iOS 9 we need to manually add the autoreleasepool.
        self.queue = DispatchQueue(label: "io.sentry.default", qos: .default)
        super.init()
    }
    
    @objc public init(queue: DispatchQueue) {
        self.queue = queue
        super.init()
    }
    
    @objc public init(utilityNamed name: String) {
        self.queue = DispatchQueue(label: name, qos: .utility)
        super.init()
    }
    
    @objc public init(conccurentUtilityNamed name: String) {
        self.queue = DispatchQueue(label: name, qos: .utility, attributes: .concurrent)
        super.init()
    }
    
    @objc public init(utilityNamed name: String, relativePriority: Int) {
        self.queue = DispatchQueue(label: name, qos: DispatchQoS(qosClass: .utility, relativePriority: relativePriority))
        super.init()
    }

    @objc(dispatchAsyncWithBlock:)
    public func dispatchAsync(_ block: @escaping () -> Void) {
        queue.async {
            autoreleasepool {
                block()
            }
        }
    }

    @objc(dispatchAsyncOnMainQueue:)
    public func dispatchAsyncOnMainQueue(block: @escaping () -> Void) {
        if Thread.isMainThread {
            block()
        } else {
            DispatchQueue.main.async {
                autoreleasepool {
                    block()
                }
            }
        }
    }
    
    @objc public func dispatchSync(block: () -> Void) {
        queue.sync(execute: block)
    }
    
    @objc(dispatchSyncOnMainQueue:)
    public func dispatchSyncOnMainQueue(block: @escaping () -> Void) {
        if Thread.isMainThread {
            block()
        } else {
            DispatchQueue.main.sync(execute: block)
        }
    }
    
    @objc(dispatchSyncOnMainQueue:timeout:)
    @discardableResult public func dispatchSyncOnMainQueue(block: @escaping () -> Void, timeout: TimeInterval) -> Bool {
        if Thread.isMainThread {
            block()
            return true
        } else {
            let semaphore = DispatchSemaphore(value: 0)
            
            DispatchQueue.main.async {
                block()
                semaphore.signal()
            }
            
            let timeoutTime = DispatchTime.now() + timeout
            return semaphore.wait(timeout: timeoutTime) == .success
        }
    }

    @objc public func dispatch(after interval: TimeInterval, block: DispatchWorkItemWrapper) {
        let delta = Int64(interval * Double(NSEC_PER_SEC))
        let when = DispatchTime.now() + DispatchTimeInterval.nanoseconds(Int(delta))
        queue.asyncAfter(deadline: when) {
            autoreleasepool {
                block.workItem.perform()
            }
        }
    }
    
    @objc
    public func dispatchCancel(_ block: DispatchWorkItemWrapper) {
        block.workItem.cancel()
    }

    @objc(createDispatchBlock:)
    public func createDispatchBlock(_ block: @escaping () -> Void) -> DispatchWorkItemWrapper? {
        return .init(workItem: DispatchWorkItem(block: block))
    }
} 

// DispatchWorkItem is not visible to objc so this class wraps it
// and is used in @objc APIs.
@_spi(Private) @objc(SentryDispatchWorkItemWrapper) public class DispatchWorkItemWrapper: NSObject {
    let workItem: DispatchWorkItem
    
    init(workItem: DispatchWorkItem) {
        self.workItem = workItem
    }
}
